api.R

```diff
@@ -1,101 +1,238 @@
+
+  # Create API -----------------------------------------------------------------
+
+  # Create model object and populate metadata
+  model <- vetiver_model(fit, "LightGBM", save_prototype = ptype)
+  model$recipe <- recipe
+  model$pv$round_type <- metadata$pv_round_type
+  model$pv$round_break <- metadata$pv_round_break[[1]]
+  model$pv$round_to_nearest <- metadata$pv_round_to_nearest[[1]]
+
+  return(model)
+}
+
+# Filter the valid runs for the run marked as default
+default_run <- valid_runs %>%
+  dplyr::filter(run_id == default_run_id) %>%
+  dplyr::slice_head()
+
+# Retrieve paths and model objects for all endpoints to be deployed
+all_endpoints <- list()
+for (i in seq_len(nrow(valid_runs))) {
+  run <- valid_runs[i, ]
+  model <- get_model_from_run(
+    run$run_id, run$year, run$dvc_bucket, run$predictors_only
   )
+  all_endpoints[[i]] <- list(
+    path = glue::glue("{base_url_prefix}/{run$run_id}"),
+    model = model,
+    is_default = run$run_id == default_run$run_id
+  )
+}
+
+# Instantiate a Plumber router for the API. Note that we have to use direct
+# Plumber calls instead of using vetiver to define the API since vetiver
+# currently has bad support for deploying multiple models on the same API
+router <- pr() %>%
+  plumber::pr_set_debug(rlang::is_interactive()) %>%
+  plumber::pr_set_serializer(plumber::serializer_unboxed_json(null = "null"))
+
+# Add Plumber POST enpdoints for each model
+for (i in seq_along(all_endpoints)) {
+  endpoint <- all_endpoints[[i]]
+  router <- plumber::pr_post(
+    router, endpoint$path, handler_predict(endpoint$model)
+  )
+  # Point the base endpoint at the default model
+  if (endpoint$is_default) {
+    router <- plumber::pr_post(
+      router, base_url_prefix, handler_predict(endpoint$model)
+    )
+  }
+}
+
+# Define a function to override the openapi spec for the API, using
+# each model's prototype for docs and examples
+modify_spec <- function(spec) {
+  spec$info$title <- "CCAO Residential AVM API"
+  spec$info$description <- (
+    "API for returning predicted values using CCAO residential AVMs"
+  )
+
+  for (i in seq_along(all_endpoints)) {
+    endpoint <- all_endpoints[[i]]
+    ptype <- endpoint$model$prototype
+    path <- endpoint$path
+    orig_post <- pluck(spec, "paths", path, "post")
+    spec$paths[[path]]$post <- list(
+      summary = glue_spec_summary(ptype),
+      requestBody = map_request_body(ptype),
+      responses = orig_post$responses
+    )
+    if (endpoint$is_default) {
+      orig_default_post <- pluck(spec, "paths", base_url_prefix, "post")
+      spec$paths[[base_url_prefix]]$post <- list(
+        summary = glue_spec_summary(ptype),
+        requestBody = map_request_body(ptype),
+        responses = orig_default_post$responses
+      )
+    }
+  }
+
+  return(spec)
+}
+
+router <- plumber::pr_set_api_spec(router, api = modify_spec) %>%
+  plumber::pr_set_docs(
+    "rapidoc",
+    header_color = "#F2C6AC",
+    primary_color = "#8C2D2D"
+  )
```

Comment:
We could definitely do more work here to align the docs site branding with CCAO guidelines, but I didn't bother at this point.

Reply:
It's unlikely that anyone except us and Derrick will look at the actual site, so I think it's fine.